packetbeat/protos/cassandra/parser.go

```diff
@@ -0,0 +1,266 @@
+package cassandra
+
+import (
+	"errors"
+	"github.com/elastic/beats/libbeat/common/streambuf"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/packetbeat/protos/applayer"
+	. "github.com/elastic/beats/packetbeat/protos/cassandra/internal/gocql"
+	"time"
+)
+
+type parser struct {
+	buf       streambuf.Buffer
+	config    *parserConfig
+	framer    *Framer
+	message   *message
+	onMessage func(m *message) error
+}
+
+type parserConfig struct {
+	maxBytes   int
+	compressor Compressor
+	ignoredOps map[FrameOp]bool
+}
+
+// check whether this ops is enabled or not
+func (p *parser) FrameOpsIgnored() bool {
+
+	if p.config.ignoredOps != nil && len(p.config.ignoredOps) > 0 {
+		//default map value is false
+		v := p.config.ignoredOps[p.framer.Header.Op]
+		if v {
+			return true
+		}
+	}
+	return false
+}
+
+type message struct {
+	applayer.Message
+
+	// indicator for parsed message being complete or requires more messages
+	// (if false) to be merged to generate full message.
+	isComplete bool
+
+	failed  bool
+	ignored bool
+
+	data   map[string]interface{}
+	header map[string]interface{}
+	// list element use by 'transactions' for correlation
+	next *message
+
+	transactionTimeout time.Duration
+
+	results transactions
+}
+
+// Error code if stream exceeds max allowed size on append.
+var (
+	ErrStreamTooLarge = errors.New("Stream data too large")
+	isDebug           = false
+)
+
+func (p *parser) init(
+	cfg *parserConfig,
+	onMessage func(*message) error,
+) {
+	*p = parser{
+		buf:       streambuf.Buffer{},
+		config:    cfg,
+		onMessage: onMessage,
+	}
+
+	isDebug = logp.IsDebug("cassandra")
+
+}
+
+func (p *parser) append(data []byte) error {
+	_, err := p.buf.Write(data)
+	if err != nil {
+		return err
+	}
+
+	if p.config.maxBytes > 0 && p.buf.Total() > p.config.maxBytes {
+		return ErrStreamTooLarge
+	}
+	return nil
+}
+
+func (p *parser) feed(ts time.Time, data []byte) error {
+	if err := p.append(data); err != nil {
+		return err
+	}
+
+	for p.buf.Total() > 0 {
+		if p.message == nil {
+			// allocate new message object to be used by parser with current timestamp
+			p.message = p.newMessage(ts)
+			p.message.data = map[string]interface{}{}
```

Comment:
why initialize p.message.data with valid map at all, why not use nil. `message.data` will be overwritten by parser anyway.